src/lib/storage/index/partial_hash/partial_hash_index_impl.cpp

```diff
@@ -0,0 +1,253 @@
+  while (map_iter != _map.end()) {
+    auto& row_ids = _map.at(map_iter->first);
+
+    // Remove every RowID entry of the value that references one of the chunks.
+    row_ids.erase(std::remove_if(row_ids.begin(), row_ids.end(), is_to_unindex), row_ids.end());
+
+    map_iter = row_ids.empty() ? _map.erase(map_iter) : ++map_iter;
+  }
+
+  auto& nulls = _null_values;
+
+  nulls.erase(std::remove_if(nulls.begin(), nulls.end(), is_to_unindex), nulls.end());
+
+  return size_before - _indexed_chunk_ids.size();
+}
+
+template <typename DataType>
+typename PartialHashIndexImpl<DataType>::IteratorPair PartialHashIndexImpl<DataType>::range_equals(
+    const AllTypeVariant& value) const {
+  const auto begin = _map.find(boost::get<DataType>(value));
+  if (begin == _map.end()) {
+    const auto end_iter = cend();
+    return std::make_pair(end_iter, end_iter);
+  }
+  auto end = begin;
+  return std::make_pair(Iterator(std::make_shared<TableIndexFlattenedSparseMapIterator<DataType>>(begin)),
+                        Iterator(std::make_shared<TableIndexFlattenedSparseMapIterator<DataType>>(++end)));
+}
+
+template <typename DataType>
+std::pair<typename PartialHashIndexImpl<DataType>::IteratorPair, typename PartialHashIndexImpl<DataType>::IteratorPair>
+PartialHashIndexImpl<DataType>::range_not_equals(const AllTypeVariant& value) const {
+  const auto range_eq = range_equals(value);
+  return std::make_pair(std::make_pair(cbegin(), range_eq.first), std::make_pair(range_eq.second, cend()));
+}
+
+template <typename DataType>
+typename PartialHashIndexImpl<DataType>::Iterator PartialHashIndexImpl<DataType>::cbegin() const {
+  return Iterator(std::make_shared<TableIndexFlattenedSparseMapIterator<DataType>>(_map.cbegin()));
+}
+
+template <typename DataType>
+typename PartialHashIndexImpl<DataType>::Iterator PartialHashIndexImpl<DataType>::cend() const {
+  return Iterator(std::make_shared<TableIndexFlattenedSparseMapIterator<DataType>>(_map.cend()));
+}
+
+template <typename DataType>
+typename PartialHashIndexImpl<DataType>::Iterator PartialHashIndexImpl<DataType>::null_cbegin() const {
+  return Iterator(std::make_shared<TableIndexVectorIterator>(_null_values.cbegin()));
+}
+
+template <typename DataType>
+typename PartialHashIndexImpl<DataType>::Iterator PartialHashIndexImpl<DataType>::null_cend() const {
+  return Iterator(std::make_shared<TableIndexVectorIterator>(_null_values.cend()));
+}
+
+template <typename DataType>
+size_t PartialHashIndexImpl<DataType>::memory_usage() const {
+  auto bytes = size_t{0u};
+
+  bytes += sizeof(_indexed_chunk_ids);  // NOLINT - Linter complains I should use vector.size(), which would be wrong.
+  bytes += sizeof(ChunkID) * _indexed_chunk_ids.size();
+
+  bytes += sizeof(_map);
+  // TODO(anyone): Find a clever way to estimate the hash sizes in the maps. For now we estimate a hash size of 8 byte
+  bytes += 8 /* hash size */ * _map.size();
+  bytes += sizeof(std::vector<RowID>) * _map.size();
```

Comment:
I think we need to have another look at this.

Both the hash map as well as the PosLists can have a capacity, that is much larger than the actual entry count.
Further, hashmaps usually (there are optimizations) don't store the hashed value but the actual key to resolve conflicts.

Reply:
I thinks it's still unresolved.

Above we add the size of the hash key. I don't understand why. The key is probably stored.
Also, something like `capacity()` should be used here, not `size()`. (or bucket count?)

Reply:
> I think we need to have another look at this.
> 
> Both the hash map as well as the PosLists can have a capacity, that is much larger than the actual entry count. Further, hashmaps usually (there are optimizations) don't store the hashed value but the actual key to resolve conflicts.



> I thinks it's still unresolved.
> 
> Above we add the size of the hash key. I don't understand why. The key is probably stored. Also, something like `capacity()` should be used here, not `size()`. (or bucket count?)

At least Tessil's sparse map does not provide a `capacity()` function:
https://github.com/Tessil/sparse-map/blob/89d30537eb5326606b6a45bbf00b4f4ff04417fa/include/tsl/sparse_map.h#L220-L225

Those functions call the following functions:
https://github.com/Tessil/sparse-map/blob/89d30537eb5326606b6a45bbf00b4f4ff04417fa/include/tsl/sparse_hash.h#L1377-L1387

> Further, hashmaps usually (there are optimizations) don't store the hashed value but the actual key

Agreed. @tjjordan Will you take care of it (affecting lines 233/234)?